src/lib/Libifl/grunt_parse.c

```diff
@@ -650,3 +652,82 @@ parse_plus_spec(char *selstr, int *rc)
 
 	return (parse_plus_spec_r(ps, &pe, &hp));
 }
+
+/**
+ * @brief
+ *	parse_plus_spec - thread safe
+ * @par
+ *	Called with "str" set for start of string of a set of plus connnected
+ *	substrings "substring1+substring2+...";  OR
+ *	called with null to continue where left off.
+ *
+ * @param[in] selstr - string holding select specs
+ * @param[in,out] tailptr - pointer to the end of the last substring
+ * @param[in] rc - flag
+ *
+ * @return 	A pointer to next substring
+ * @retval	next substring (char *)
+ * @retval	NULL if end of the spec
+ *
+ * @par
+ *	IMPORTANT: the "selstr" is copied into a locally allocated
+ *	char array for parsing.  The orignal string is untouched.  The array
+ *	is grown as need to hold "selstr".
+ */
+char *
+parse_plus_spec_mt_safe(char *selstr, char **tailptr, int *rc)
```

Comment:
hmm .. so parse_plus_spec_mt_safe() isn't similar parse_plus_spec_r(), it's actually similar to parse_plus_spec() instead. I see both parse_plus_spec() and parse_plus_spec_mt_safe() as convenience wrapper functions around parse_plus_spec_r() as they both call parse_plus_spec_r(). So, if I change parse_plus_spec_r() to also do the job that parse_plus_spec_mt_safe() does, then we'll have to create a new function which does what parse_plus_spec_r() currently does so that parse_plus_spec() can call that new function instead of parse_plus_spec_r(). Maybe it makes more sense to rename pbs_plus_spec_r() to  pbs_plus_spec_2() ? I think this is just a naming issue. Let me know what you think.

Reply:
Like I said above, I think you should make one function that is MT safe.

Reply:
But that's what I have right now. I needed a MT safe version of parse_plus_spec(), which is  what parse_plus_spec_mt_safe() is. parse_plus_spec_mt_safe() and parse_plus_spec() both call parse_plus_spec_r(), so parse_plus_spec_r() is a lower level function, that's why I suggested renaming it to parse_plus_spec_2().

Reply:
No, you are creating a second function that does exactly what parse_plus_spec() does, but is MT safe.  What I am suggesting is making parse_plus_spec() MT safe, so we only have one version of the function.

Reply:
Oh ok. But won't that make things slower for other daemons that call parse_plus_spec()? It will cause memory to be allocated on the heap every time instead of just once (https://stackoverflow.com/questions/2264969/why-is-memory-allocation-on-heap-much-slower-than-on-stack)

Reply:
I'm not talking about allocating memory on the heap.  I'm talking abut turning parse_plus_spec() into something like strtok_r().  You pass an opaque pointer into the call for the function to use.